packages/button/src/index.tsx

```diff
@@ -32,55 +38,103 @@ const StyledText = styled(Text)<StyledTextProps>`
         }
       : {
           display: 'inline-flex'
         }}
   ${_ => getButtonStyle(_.appearance, _.intent)};
+  ${_ => (_.icon ? getIconAttachmentStyle(_.iconPosition) : {})};
+  ${SvgWrapper} {
+    display: flex;
+    align-items: center;
+    margin-left: ${_ =>
+      _.iconPosition === 'center'
+        ? 0
+        : _.iconPosition === 'left'
+        ? `${-Math.round(_.height / 8)}px`
+        : `${Math.round(_.height / 4)}px`};
+    margin-right: ${_ =>
+      _.iconPosition === 'center'
+        ? 0
+        : _.iconPosition === 'left'
+        ? `${Math.round(_.height / 4)}px`
+        : `${-Math.round(_.height / 8)}px`};
+  }
 `
 const StyledSpinner = styled(Spinner)<StyledSpinnerProps>`
   opacity: 0.6;
   color: ${_ => _.theme.scales.neutral.N7A};
-  margin-right: ${_ => _.marginRight}px;
-  margin-left: ${_ => _.marginLeft}px;
+  margin-right: ${_ =>
+    _.iconPosition === 'center'
+      ? 0
+      : _.iconPosition === 'left'
+      ? Math.round(_.height / 4)
+      : -Math.round(_.height / 8)}px;
+  margin-left: ${_ =>
+    _.iconPosition === 'center'
+      ? 0
+      : _.iconPosition === 'left'
+      ? -Math.round(_.height / 8)
+      : Math.round(_.height / 4)}px;
   vertical-align: middle;
 `
 
 const ButtonFCFactory: <AdditionalProps extends {}>(
   as: StyledComponentElement
 ) => React.FC<ButtonProps> = <AdditionalProps extends {}>(
   as: StyledComponentElement = 'button'
-) => ({children, innerRef, ...props}) => {
-  // eslint-disable-next-line react-hooks/rules-of-hooks
-  const defaults = useDefaults('button', props, {
-    height: 32,
-    appearance: 'default' as ButtonAppearanceType,
-    intent: 'none' as ButtonIntentType,
-    isLoading: false,
-    full: false
-  })
+) => {
+  const ButtonFC = ({children, innerRef, ...props}) => {
+    const defaults = useDefaults('button', props, {
+      height: 32,
+      appearance: 'default' as ButtonAppearanceType,
+      intent: 'none' as ButtonIntentType,
+      isLoading: false,
+      full: false,
+      iconPosition: 'right' as ButtonIconPosition
+    })
+
+    const theme = useTheme()
+
+    const IconComponent = props.icon
+
+    return (
+      <StyledText
+        as={as}
+        borderRadius={getBorderRadiusForControlHeight(defaults.height)}
+        variant={getTextSizeForControlHeight(defaults.height)}
+        ref={innerRef}
+        {...defaults}
+        {...props}
+        disabled={props.disabled || props.isLoading}
+      >
+        {children}
+        {defaults.isLoading && (
+          <StyledSpinner
+            iconPosition={props.iconPosition}
+            height={defaults.height}
+            size={Math.round(defaults.height / 2)}
+          />
+        )}
+        {IconComponent && !defaults.isLoading && (
+          <SvgWrapper>
+            <IconComponent
+              size={defaults.height / 2}
+              color={
+                props.appearance === 'primary'
+                  ? 'white'
```

Comment:
1) It doesn't take `disabled` prop in consideration
2) Instead of hardcoding colors here, try to reuse the same logic as used in `getButtonStyle`. Keep in mind that we shouldn't keep any values in more than a one place so future changes won't be bug-prone. You can abstract/extract text colors in `getButtonStyle` to an external function and then use it here.